File: 3106-lexicographically-smallest-string-after-operations-with-constraint/3106-lexicographically-smallest-string-after-operations-with-constraint.cpp
class Solution {
public:
    string getSmallestString(string s, int k) {
        string smallestString = "";
        int n = s.size(), curr = 0;
        while(curr < n){
            if(k > 0){
                int distance = s[curr] - 'a';
                if( (26 - distance < distance ) && k >= (26 - distance) ){
                    k -= 26 - distance;
                    smallestString += 'a';
                }else{
                    int temp = min(k, distance);
                    k -= temp;
                    smallestString += s[curr] - temp;
                }
            }else{
                smallestString += s[curr];
            }
            curr++;
        }
        return smallestString;
    }
};

File: 3106-lexicographically-smallest-string-after-operations-with-constraint/3106-lexicographically-smallest-string-after-operations-with-constraint.py
class Solution:
    def getSmallestString(self, s: str, k: int) -> str:
        if k==0:
            return s
        i = 0
        while k>=0 and i<len(s):
            tmp = ord(s[i])-ord('a')
            tmp = min(tmp, 26-tmp)
            if tmp<=k:
                s = s[:i]+'a'+s[i+1:]
                k -= tmp
                i += 1
            else:
                s = s[:i]+chr(ord(s[i])-k)+s[i+1:]
                break
        return s